tools/umbra_layout.py

```diff
@@ -0,0 +1,566 @@
+'''
+ **********************************************************
+ * Copyright (c) 2020 Google, Inc.  All rights reserved.
+ * ********************************************************
+'''
+'''
+  Redistribution and use in source and binary forms, with or without
+  modification, are permitted provided that the following conditions are met:
+
+ * Redistributions of source code must retain the above copyright notice,
+   this list of conditions and the following disclaimer.
+
+ * Redistributions in binary form must reproduce the above copyright notice,
+   this list of conditions and the following disclaimer in the documentation
+   and/or other materials provided with the distribution.
+
+ * Neither the name of Google, Inc. nor the names of its contributors may be
+   used to endorse or promote products derived from this software without
+   specific prior written permission.
+
+ THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ ARE DISCLAIMED. IN NO EVENT SHALL GOOGLE, INC. OR CONTRIBUTORS BE LIABLE
+ FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+ DAMAGE.
+'''
+'''
+A script used to find Umbra Shadow Memory Layouts.
+To format script with yapf: yapf -i --style='{COLUMN_LIMIT: 90}' umbra_layout.py
+'''
+
+import sys
+import argparse
+from enum import Enum
+from z3 import *
+
+PTR_SIZE = 64
+
+
+def get_formatted_hex(integer):
+    # Returns a formatted hex string.
+    return '0x{0:0{1}X}'.format(integer, 16)
+
+
+class OS(Enum):
+    LINUX = 'linux'
+    WINDOWS8 = 'windows8'
+    WINDOWS81 = 'windows81'
+
+    def __str__(self):
+        return self.value
+
+    def get_app_regions(os):
+        # Returns pre-defined app regions of the passed Operating System.
+        regions = []
+        if os == OS.LINUX:
+            regions.append(Region('exec,heap, data', 0x0, 0x10000000000))
+            regions.append(Region('pie', 0x550000000000, 0x570000000000))
+            regions.append(Region('lib, map, stack, vdso', 0x7F0000000000,
+                                  0x800000000000))
+
+            # FIXME: Should we map shadow memory for vsyscall?
+            # Doing so can prevent the possibility of a SAT layout.
```

Comment:
But we have to, and the current C code handles vsyscall.  Are you saying there's only code and no data and nobody shadows code?

Reply:
Yeah, back then I was thinking of whether we need to have shadow for code. However, "we have" to because we want to catch erroneous use of memory right?

Reply:
Right, I think we have to handle all mapped app memory regions.

Reply:
Done.

Reply:
Hmm, so I was doing some final checks before I go ahead with the merge, and I note that if we include vsyscall, then certain layouts are not possible.

Currently, on Linux, if we select scale x2 up, we get a collision. 

```
*** Umbra Shadow Memory Layout ***

OS: linux 

Memory Layout:
	Mask: 0x00000FFFFFFFFFFF
	Disp: 0x0000048000000000
	Unit Size: 0x0000100000000000
	Scale UP: True
	Scale: 1


Collision: 0x000028FFFEC00000 0x000028FFFFFFFFFE
FAILED

MAP 0
	app0:	 [0x0000000000000000, 0x000000FFFFFFFFFF] : exec,heap, data
	shd0:	 [0x0000090000000000, 0x00000AFFFFFFFFFE]


	app1:	 [0x0000550000000000, 0x000055FFFFFFFFFF] : pie
	shd1:	 [0x0000130000000000, 0x000014FFFFFFFFFE]


	app2:	 [0x00007F0000000000, 0x00007FFFFFFFFFFF] : lib, map, stack, vdso
	shd2:	 [0x0000270000000000, 0x000028FFFFFFFFFE]


	app3:	 [0xFFFFFFFFFF600000, 0xFFFFFFFFFF600FFF] : vsyscall
	shd3:	 [0x000028FFFEC00000, 0x000028FFFEC01FFE]

```

It collides with lib region. This is the layout setting we use today. It also appears that the problem is only when we scale up shadow memory (and not scale down).

Reply:
https://github.com/DynamoRIO/drmemory/blob/ef8db941a3db47e4a6778448be68f53a8db048a0/umbra/umbra_64.c#L314

Reply:
If I try to get a solution from the model, I get unsat:

```
$ python3 umbra_layout.py --os linux --mask 0xFFFFFFFFFFF --scale up_2x --verify --count 1
*** Umbra Shadow Memory Layout ***

OS: linux 

FAILED

```

Reply:
As your link above points at, the code today has a custom disp for up-2x.  Are you able to add varying disps per scale to your model?

Reply:
Yeah, it does. The script can check fixed disps passed as param, or ask the solver to provide custom disps (if sat).

I checked that the disp currently used by Umbra here: https://github.com/DynamoRIO/drmemory/pull/2300#discussion_r537068124

I asked the solver to provide me with a custom disp (if sat),  here: https://github.com/DynamoRIO/drmemory/pull/2300#discussion_r537069155

But it's not sat.

Reply:
So you're saying that the current Umbra code is incorrect.  But I thought there were sanity checks inside the umbra code itself, plus tests like umbra_all_scales: are there not enough checks there, or is there a bug in one of them?

Reply:
Not a bug per se because Umbra code ignores vsyscall.

https://github.com/DynamoRIO/drmemory/blob/55a9c58b75ab5dbcabdb6368c78132d2eab6fbc1/umbra/umbra_64.c#L357

If however, we consider it in our model, then it is unsat. I guess the main point is that Umbra code is currently ignoring mapping of vsyscall and thus breaking its contract.

https://github.com/DynamoRIO/drmemory/pull/2300#discussion_r532875525

Reply:
Perhap, we shoudl revert, adding again the FIXME comment?

Reply:
I'm still confused: I don't see the Umbra code deliberately ignoring vsyscall?
For what you're quoting:
```
    /* We split app3 [0x7F0000000000, 0x800000000000) into two parts:
```
It goes on to say:
```
     * [0x7F0000000000, 0x7FFFFF400000) and [0x7FFFFF800000, 0x800000000000).
     * And we skip [0x7FFFFF400000-0x7FFFFF800000)
...
     * We assume [0x7FFFFF400000-0x7FFFFF800000) will not be used by app.
     * If app allocates memory from that region, umbra_add_app_segment
     * will fail because umbra_add_shadow_segment fails to add corresponding
     * shadow memory segment.
```

So if a vsyscall were to be in that skipped region, it would just fail up front.
Is that what you mean?  That's not great, but it's not what I thought you were
saying, which is that it pretending to have a correct mapping.

Reply:
Hmm, basically, this thread concerns whether the layout should consider the shadow of vsyscall. Initially, the .py script did not consider this region and included a FIXME comment, highlighting that some layouts won't be possible otherwise. After some discussion, we removed this comment, and made the .py script to also take into account the vsyscall region.

However, the model does not skip any part of the app3. Consequently, a collision is detected.

I guess my mistake in explaining this was confusing the skipped region with vsyscall but I think my overall issue is still there nonetheless. By not considering mapping shadow to the skipped region, aren't we breaking Umbra's contract?

Reply:
There are possible remediations: If the kernel isn't the one placing something in the gap at process init (like vsyscall) and it's the app who does an explicit mmap there or ld.so tries to load a library there, umbra can reject the mmap: so it's ok to have a gap if you can enforce it (though it's not today: #1798).  Or umbra could handle an unmovable alloc there by using expensive mappings: I think that's what #1782 proposes and the comment there implies.

Reply:
I'm also confused by the vsyscall address in this PR: isn't vsyscall subject to ASLR and it moves around?

Reply:
For the problematic app region where umbra_64.c adds a gap: wouldn't other tools who do simple displacement shadow mapping also hit a problem?  E.g., AddressSanitizer.

Reply:
> I'm also confused by the vsyscall address in this PR: isn't vsyscall subject to ASLR and it moves around?

So if we're fairly certain vsyscall has nothing but code wrappers and no data, is not subject to ASLR so we know where it is, and is always separate from the larger and randomized vdso and vvar etc. which we are shadowing: then maybe skipping vsyscall is ok?

Reply:
> I'm also confused by the vsyscall address in this PR: isn't vsyscall subject to ASLR and it moves around?

Isn't the location of vsyscall fixed? hence the introduction of vdso?

> For the problematic app region where umbra_64.c adds a gap: wouldn't other tools who do simple displacement shadow mapping also hit a problem? E.g., AddressSanitizer.

This is a good question. I do not know how Asan handles such issues, but it would be interesting to know. I'll try and search.

>  maybe skipping vsyscall is ok?

That is what I was thinking initially. I guess we need to decide whether we are going to change .py or Umbra code (i.e., skip a region in app3 or skip vsyscall), because we want the implementation to match the model.

@zhaoqin Any opinion on the above please?

Reply:
I had some time to continue on this issue. I had a quick look at how ASAN sets up and maps shadow memory.

Doc:  https://github.com/google/sanitizers/wiki/AddressSanitizerAlgorithm
Src: https://code.woboq.org/llvm/llvm/lib/Transforms/Instrumentation/AddressSanitizer.cpp.html

According to ASAN's [mapping](https://code.woboq.org/llvm/compiler-rt/lib/asan/asan_mapping.h.html#21),  0x00007fff8000 - 0x10007fff7fff is reserved for shadow (and shadow's shadow).

```
// Typical shadow mapping on Linux/x86_64 with SHADOW_OFFSET == 0x00007fff8000:
--
// \|\| `[0x10007fff8000, 0x7fffffffffff]` \|\| HighMem    \|\|
// \|\| `[0x02008fff7000, 0x10007fff7fff]` \|\| HighShadow \|\|
// \|\| `[0x00008fff7000, 0x02008fff6fff]` \|\| ShadowGap  \|\|
// \|\| `[0x00007fff8000, 0x00008fff6fff]` \|\| LowShadow  \|\|
// \|\| `[0x000000000000, 0x00007fff7fff]` \|\| LowMem     \|\|
//
```

The shadow memory block is reserved upon initialization:  https://code.woboq.org/llvm/compiler-rt/lib/asan/asan_shadow_setup.cpp.html#126

Just for context, one thing to note is that ASAN maps an app byte to 1 bit - working at the granularity of 8 bytes. In contrast, Dr. Memory shadows two bits per byte, as denoted [here](https://github.com/DynamoRIO/drmemory/blob/04b22efd7e22f319c1c6e90f40186ec18500ad64/drmemory/shadow.c#L170).

Reply:
Perhaps, we should do the model like how Umbra does it today, i.e., skip a region in app3. The issue would then be resolved with https://github.com/DynamoRIO/drmemory/issues/1798 where Umbra would enforce that this region is not used by the app.

Reply:
@derekbruening I pushed the region split. Do you think this is good please?

Reply:
Trying to context switch back in here: so this .py is now matching Umbra, and we just need to get Umbra to actually enforce its skip there.  Yes, SGTM.

Reply:
Great, thanks. We are on the same page.